polkadot/erasure-coding/src/lib.rs

```diff
@@ -113,6 +121,66 @@ fn code_params(n_validators: usize) -> Result<CodeParams, Error> {
 	})
 }
 
+/// Reconstruct the v1 available data from the set of systematic chunks.
+///
+/// Provide a vector containing chunk data. If too few chunks are provided, recovery is not
+/// possible.
+pub fn reconstruct_from_systematic_v1(
+	n_validators: usize,
+	chunks: Vec<&[u8]>,
+) -> Result<AvailableData, Error> {
+	reconstruct_from_systematic(n_validators, chunks)
+}
+
+/// Reconstruct the available data from the set of systematic chunks.
+///
+/// Provide a vector containing chunk data. If too few chunks are provided, recovery is not
+/// possible.
+pub fn reconstruct_from_systematic<T: Decode>(
+	n_validators: usize,
+	chunks: Vec<&[u8]>,
+) -> Result<T, Error> {
+	let kpow2 = systematic_recovery_threshold(n_validators)?;
+
+	let Some(first_shard) = chunks.iter().next() else { return Err(Error::NotEnoughChunks) };
+	let shard_len = first_shard.len();
+
+	if shard_len == 0 {
+		return Err(Error::NonUniformChunks)
+	}
+
+	if shard_len % 2 != 0 {
+		return Err(Error::UnevenLength)
+	}
+
+	if chunks.len() < kpow2 {
+		return Err(Error::NotEnoughChunks)
+	}
+
+	let mut check_shard_len = true;
+	let mut systematic_bytes = Vec::with_capacity(shard_len * kpow2);
+
+	for i in (0..shard_len).step_by(2) {
+		for chunk in chunks.iter().take(kpow2) {
+			if check_shard_len {
+				if chunk.len() != shard_len {
+					return Err(Error::NonUniformChunks)
+				}
+			}
+
+			// No need to check for index out of bounds because i goes up to shard_len and
+			// we return an error for non uniform chunks.
+			systematic_bytes.push(chunk[i]);
+			systematic_bytes.push(chunk[i + 1]);
+		}
+
+		// After the first check, stop checking the shard lengths.
+		check_shard_len = false;
```

Comment:
why? a malicious "erasure coding" can make the nodes panic with oob access?

Reply:
we iterate through the chunks vector multiple times, so I added this flag to not check the chunk lengths on every iteration, but only on the first one (since they don't change on subsequent iterations). So yes, not performing this check would result in an OOB panic, given the current code

having chunks with unfiform lengths is just an invariant of the encoding scheme.

Reply:
I thought about this but I wanted to minimise the number of passes we do over the vector.
Still, LLVM should be smart enough to optimise this, so I'll apply your suggestion